trino/client.py

```diff
@@ -925,3 +977,217 @@ def _parse_retry_after_header(retry_after):
+        super().__init__(segment)
+        self._segment = cast(_InlineSegmentTO, segment)
+        self._row_mapper = row_mapper
+
+    @property
+    def rows(self) -> List[List[Any]]:
+        inline_segment = self._segment
+        data = inline_segment["data"]
+        decoded_string = base64.b64decode(data)
+        rows = self._row_mapper.map(json.loads(decoded_string))
+        return list(iter(rows))
+
+    def __repr__(self):
+        return f"InlineSegment(metadata={self.metadata})"
+
+
+class SpooledSegment(Segment):
+    """
+    A subclass of Segment that handles spooled data segments, where data may be compressed and needs to be
+    retrieved via HTTP requests. The segment includes methods for acknowledging processing and loading the
+    segment from remote storage.
+
+    Attributes:
+        rows (property): The rows of data, loaded and mapped from the spooled segment.
+        uri (property): The URI for the spooled segment.
+        ack_uri (property): The URI for acknowledging the processing of the spooled segment.
+        headers (property): The headers associated with the spooled segment.
+
+    Methods:
+        acknowledge(): Sends an acknowledgment request for the segment.
+    """
+    def __init__(
+        self,
+        response: _SpooledProtocolResponseTO,
+        segment: _SpooledSegmentTO,
+        row_mapper: RowMapper,
+        request: TrinoRequest
+    ) -> None:
+        super().__init__(segment)
+        self._segment = cast(_SpooledSegmentTO, segment)
+        self._row_mapper = row_mapper
+        self._request = request
+        self._encoding = response["encoding"]
+        self._headers = segment.get("headers")
+
+    @property
+    def rows(self) -> List[List[Any]]:
+        return self._row_mapper.map(json.loads(self._load_spooled_segment()))
+
+    @property
+    def uri(self) -> str:
+        return self._segment["uri"]
+
+    @property
+    def ack_uri(self) -> str:
+        return self._segment["ackUri"]
+
+    @property
+    def headers(self) -> Dict[str, List[str]]:
+        return self._segment.get("headers", {})
+
+    def acknowledge(self):
+        http_response = self._send_spooling_request(self._segment["ackUri"], self._segment)
+        if not http_response.ok:
+            self._request.raise_response_error(http_response)
+
+    def _load_spooled_segment(self) -> str:
+        segment = self._segment
+        http_response = self._send_spooling_request(segment["uri"], segment)
+        if not http_response.ok:
+            self._request.raise_response_error(http_response)
+        content = http_response.content
+
+        encoding = self._encoding
+        if encoding == "json+zstd":
+            zstd_decompressor = zstandard.ZstdDecompressor()
+            return zstd_decompressor.decompress(content).decode('utf-8')
+        elif encoding == "json+lz4":
+            expected_size = segment["metadata"]["uncompressedSize"]
```

Comment:
not every json+lz4 is compressed, if the data is below certain threshold, it can stay uncompressed. the algorithm is:

```
if exists metadata["uncompressedSize"]:
   decompress(data)
else
   data
```

you should also validate whether the data after decompression is `uncompressedSize`
and whether the retrieved data is `segmentSize`